### DISCUSSION TITLE
Is there any way to debug lua code?

### DISCUSSION THREAD

**Discussion by jazzcake** [Fumohouse/godot-luau-script#2] [General]
Hello,

I have just used this project and I would like to use a breakpoint in my Lua script. However, it doesn't seem to be working in the Godot script editor.

Am I missing something? Everything has been compiled with debug symbols, including the Godot engine. Is there any way to make this work or does it require additional development?

I would appreciate any feedback. Thank you.

**Comment by voidedWarranties** [comment]
Hi! Thanks for your interest.

Debugging is one of the prominent things still missing from this project, and as of yet I'm not sure how to implement in-editor debugging or whether it's even currently possible in GDExtension/with Lua (the way debugging is implemented Godot-side seems fairly GDScript-centric).

For now, you will have to use `print`. Sorry for the inconvenience.

By the way, debug symbols are only needed for debugging machine code (i.e. the stuff written in C++). You won't need it for debugging Lua, whenever that is implemented.

**Reply by jazzcake** [reply]
Thank you for your feedback,

For me, this is a must-have feature. So I'll approach this by understanding how typical Lua debugger and GDScript implementation in Godot. I'll leave some comments about progress from time to time. (Also with some question....)

**Reply by voidedWarranties** [reply]
After doing more research, I think I understand the Luau and Godot systems enough to find a path to implementing the debugger.
It will be blocked on some Godot changes which I am currently working on; I'll probably implement the debugger using these changes before PRing them to Godot `master`.
I believe it will be possible to get this in by Godot 4.1 (or its pre-releases) - the chances of merging such a change in 4.0 is quite slim. No guarantees though.

Thanks for offering your help.

**Reply by jazzcake** [reply]
Before making your implementation, can I hear your idea for path? I'm newbie on Godot, so I'm troubling with source code of Godot now.

**Reply by voidedWarranties** [reply]
I already made a working test implementation a few days ago and submitted the PR to Godot (https://github.com/godotengine/godot/pull/73967, which is in the 4.x milestone).
Here are some of the details:

* Godot handles script debugging with `EngineDebugger` (`core/debugger/engine_debugger.h`) and `ScriptDebugger` (`core/debugger/script_debugger.h`).
* `ScriptDebugger::debug` is called by the language whenever it wants to trigger a debug break. This method calls into `EngineDebugger::debug` (a virtual method implemented by the 2 debugger types) which will block the thread with a while loop and listen for commands from the editor.
* By the way Luau works, the `debug` method blocking the thread is enough to stop (most) Luau execution.
* Debug information can be found through `lua_getinfo` and is exposed to Godot through a series of `ScriptLanguageExtension` methods like `_debug_get_error`.
* `ScriptDebugger` maintains `lines_left` and `depth` (exposed through setters/getters) to indicate when execution should continue or stop again. Once set, the language has to keep track of these to determine when to break.
    * `lines_left` is the number of lines that should execute before the language should trigger a debug break again.  We can track line execution through `lua_singlestep` and the `debugstep` callback combined with `lua_getinfo`. Godot sets `lines_left` to -1 when execution can continue, and to a positive number to step through the code.
    * `depth` is the current function execution depth relative to the function in which the debugger was triggered. Depth can be tracked through `lua_stackdepth`. Godot sets `depth` to -1 when the language should decrement `lines_left` on every line, and to 0 when the language should track depth and only decrement `lines_left` if `depth` is 0 (i.e. to stay in the same function).
* `ScriptDebugger` maintains a map of breakpoints exposed through `get_breakpoints`. Currently, Luau has to poll this every frame to update breakpoints on scripts through `lua_breakpoint`.

Some implementation details are left out, but this is much of the confusing stuff. Looking through the classes I mentioned, GDScript's debugger implementation ([here](https://github.com/godotengine/godot/blob/0bc36c8954328a36295d50dc5f7722c9463f6f5c/modules/gdscript/gdscript_vm.cpp#L3526-L3548)), and the `EngineDebugger` implementations (`core/debugger/remote_debugger.cpp`, `core/debugger/local_debugger.cpp`) should give you more information. On the Luau side,  you can check `ldebug.cpp` and `lua.h` for key functions and structures (`lua_Callbacks` and `lua_Debug`). You could also check my PR which contains some documentation on `EngineDebugger`/`ScriptDebugger`.

As the Godot PR will probably take some time to be considered, I won't spend much more time on the debugger until it's merged. I might put up a branch later with the debugger functionality, but you will need to compile Godot yourself to use it.

**Reply by jazzcake** [reply]
Thank you for sharing all details. They help me a lot to understand "the confusing stuffs". 👍 Thanks!
